include/graph_contracts.hpp

```diff
@@ -0,0 +1,444 @@
+
+    [[nodiscard]] constexpr supported_type pmt_type() const noexcept { return T(); }
+        [[nodiscard]] constexpr std::string_view name() const noexcept {
+            if constexpr (PortName.empty()) {
+                return static_cast<std::string_view>(PortName);
+            } else {
+                return _port_name;
+            }
+        }
+
+        [[nodiscard]] constexpr static bool optional() noexcept { return OPTIONAL; }
+        [[nodiscard]] constexpr std::int16_t priority() const noexcept { return _priority; }
+
+        [[nodiscard]] constexpr static std::size_t available() noexcept { return 0; } //  ↔ maps to Buffer::Buffer[Reader, Writer].available()
+        [[nodiscard]] constexpr std::size_t n_history() const noexcept {
+            if constexpr (N_HISTORY == std::dynamic_extent) {
+                return _n_history;
+            } else {
+                return N_HISTORY;
+            }
+        }
+        [[nodiscard]] constexpr std::size_t min_buffer_size() const noexcept {
+            if constexpr (MIN_SAMPLES == std::dynamic_extent) {
+                return _min_samples;
+            } else {
+                return MIN_SAMPLES;
+            }
+        }
+        [[nodiscard]] constexpr std::size_t max_buffer_size() const noexcept {
+            if constexpr (MAX_SAMPLES == std::dynamic_extent) {
+                return _max_samples;
+            } else {
+                return MAX_SAMPLES;
+            }
+        }
+
+        [[nodiscard]] constexpr connection_result_t resize_buffer(std::size_t min_size) noexcept {
+            if constexpr (IS_INPUT) {
+                static_assert(IS_INPUT, "can only resize output buffers");
+            } else {
+                try {
+                    if constexpr (IS_INPUT) {
+                        _ioHandler = BufferType(min_size).new_reader();
+                        _connected = true;
+                    } else { _ioHandler =  BufferType(min_size).new_writer(); }
+                } catch (...) {
+                    return connection_result_t::FAILED;
+                }
+            }
+            return connection_result_t::SUCCESS;
+        }
+        [[nodiscard]] BufferType get_buffer() { return _ioHandler.get_buffer(); }
+
+        void setBuffer(gr::Buffer auto buffer) {
+            if constexpr (IS_INPUT) {
+                _ioHandler = std::move(buffer.new_reader());
+                _connected = true;
+            } else { _ioHandler = std::move(buffer.new_writer()); }
+        }
+
+        [[nodiscard]] ReaderType& reader() noexcept {
+            if constexpr (PortDirection == port_direction_t::OUTPUT) {
+                static_assert(helper::always_false<T>, "reader() not applicable for outputs (yet)");
+            }
+            return _ioHandler;
+        }
+
+        [[nodiscard]] WriterType& writer() noexcept {
+            if constexpr (PortDirection == port_direction_t::INPUT) {
```

Comment:
Wrong condition (acc to the static assert below)

Reply:
fixed the string description